quaint/src/connector/postgres/native/websocket.rs

```diff
@@ -95,3 +109,121 @@ impl From<TungsteniteError> for error::Error {
+        }
+    }
+}
+
+impl WsBytesStream {
+    fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut WebSocketStream<MaybeTlsStream<TcpStream>>> {
+        self.project().0
+    }
+}
+
+impl AsyncRead for WsTunnel {
+    fn poll_read(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut ReadBuf<'_>) -> Poll<std::io::Result<()>> {
+        self.project().inner.poll_read(cx, buf)
+    }
+}
+
+impl AsyncBufRead for WsTunnel {
+    fn poll_fill_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<&[u8]>> {
+        self.project().inner.poll_fill_buf(cx)
+    }
+
+    fn consume(self: Pin<&mut Self>, amt: usize) {
+        self.project().inner.consume(amt)
+    }
+}
+
+impl AsyncWrite for WsTunnel {
+    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<std::io::Result<usize>> {
+        let this = self.get_mut();
+        let sink = &mut this.inner.get_mut().0;
+        let to_io_err = |err| IoError::new(IoErrorKind::Other, err);
+
+        match this.write_state {
+            WriteState::Free => {
+                ready!(sink.poll_ready_unpin(cx)).map_err(to_io_err)?;
+                sink.start_send_unpin(Message::Binary(Bytes::copy_from_slice(buf)))
+                    .map_err(to_io_err)?;
+                this.write_state = WriteState::Writing(buf.len());
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+            WriteState::Writing(len) => {
+                ready!(sink.poll_flush_unpin(cx)).map_err(to_io_err)?;
+                this.write_state = WriteState::Free;
+                Poll::Ready(Ok(len))
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        self.project()
+            .inner
+            .get_pin_mut()
+            .get_pin_mut()
+            .poll_flush(cx)
+            .map_err(|err| IoError::new(IoErrorKind::Other, err))
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<std::io::Result<()>> {
+        self.project()
+            .inner
+            .get_pin_mut()
+            .get_pin_mut()
+            .poll_close(cx)
+            .map_err(|err| IoError::new(IoErrorKind::Other, err))
+    }
+}
+
+impl Stream for WsBytesStream {
+    type Item = Result<Bytes, IoError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.get_pin_mut().poll_next(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Ready(Some(Ok(msg))) => match msg {
+                Message::Binary(data) => Poll::Ready(Some(Ok(data))),
+                Message::Close(_) => Poll::Ready(None),
+                Message::Text(data) => {
+                    tracing::warn!(%data, "unexpected text frame in a WebSocket tunnel");
+                    cx.waker().wake_by_ref();
+                    Poll::Pending
+                }
```

Comment:
Shouln't this result in an error?

Reply:
No, that was what my initial implementation did but I changed it because in the future we might want to use text frames to send our custom packets intermingled with the tunneled stream in binary frames, and it would make it a breaking change if we returned an error here.

Reply:
see this slack thread: https://prisma-company.slack.com/archives/C04EPL7N2KV/p1737037434945399